src/Workspaces/Core/Portable/Workspace/EventMap.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis;
+
+internal sealed class EventMap
+{
+    private readonly SemaphoreSlim _guard = new(initialCount: 1);
+    private readonly Dictionary<string, EventHandlerSet> _eventNameToHandlerSet = [];
+
+    public WorkspaceEventRegistration AddEventHandler(string eventName, WorkspaceEventHandlerAndOptions handlerAndOptions)
+    {
+        using (_guard.DisposableWait())
+        {
+            _eventNameToHandlerSet[eventName] = _eventNameToHandlerSet.TryGetValue(eventName, out var handlers)
+                ? handlers.AddHandler(handlerAndOptions)
+                : EventHandlerSet.Empty.AddHandler(handlerAndOptions);
+        }
+
+        return new WorkspaceEventRegistration(this, eventName, handlerAndOptions);
+    }
+
+    public void RemoveEventHandler(string eventName, WorkspaceEventHandlerAndOptions handlerAndOptions)
+    {
+        using (_guard.DisposableWait())
+        {
+            if (_eventNameToHandlerSet.TryGetValue(eventName, out var handlers))
```

Comment:
would not finding this indicate a bug?  should we at least debug assert here?

Reply:
I feel ok adding this now that WorkspaceEventRegistration.Dispose was modified to be safe against double calls.